src/Icicle/Avalanche/FromCore.hs

```diff
@@ -183,20 +184,36 @@ insertStream namer inputType strs reds (n, strm)
 
        -- If within i days
        CS.SWindow _ newerThan olderThan inp
-        -> let factDate  = namerElemPrefix namer (namerDate namer)
-               nowDate   = namerDate namer
-               diff      = xPrim (PrimMinimal $ Min.PrimDateTime Min.PrimDateTimeDaysDifference)
+        -> let
+               genDiff :: WindowUnit -> (X.Exp () n, X.Exp () n)
+               genDiff (Days   d) = (constI d     ,xPrim (PrimMinimal $ Min.PrimDateTime Min.PrimDateTimeMinusDays))
+               genDiff (Weeks  w) = (constI (7*w) ,xPrim (PrimMinimal $ Min.PrimDateTime Min.PrimDateTimeMinusDays))
+               genDiff (Months m) = (constI m     ,xPrim (PrimMinimal $ Min.PrimDateTime Min.PrimDateTimeMinusMonths))
 
-               check  | Just o' <- olderThan
+               (>~~) = prim2 (PrimMinimal $ Min.PrimRelation Min.PrimRelationGt DateTimeT)
+               infix 4 >~~
+
+               (>=~~) = prim2 (PrimMinimal $ Min.PrimRelation Min.PrimRelationGe DateTimeT)
+               infix 4 >=~~
+
+               (<=~~) = prim2 (PrimMinimal $ Min.PrimRelation Min.PrimRelationLe DateTimeT)
+               infix 4 <=~~
+
```

Comment:
Yeah, it's the ugliest bit. Got it working, need to make it not terrible before I merge.